libraries/AP_Math/control.cpp

```diff
@@ -258,63 +228,70 @@ void shape_vel_accel_xy(const Vector2f &vel_input1, const Vector2f& accel_input,
         return;
     }
 
     // Calculate time constants and limits to ensure stable operation
-    const float KPv = 1.0 / (CONTROL_TIME_CONSTANT_RATIO*tc);
-    const float accel_tc_max = accel_max*(1-1.0f/CONTROL_TIME_CONSTANT_RATIO);
+    const float KPv = jerk_max / (CONTROL_TIME_CONSTANT_RATIO * MAX(-accel_min, accel_max));
+    const float accel_tc_max = MIN(-accel_min, accel_max) * (1.0 - 1.0 / CONTROL_TIME_CONSTANT_RATIO);
 
     // position error to be corrected
     float pos_error = pos_input - pos;
 
     // velocity to correct position
     float vel_target = sqrt_controller(pos_error, KPv, accel_tc_max, dt);
 
-    // limit velocity correction to vel_correction_max
-    if (is_positive(vel_correction_max)) {
-        vel_target = constrain_float(vel_target, -vel_correction_max, vel_correction_max);
+    // limit velocity to vel_max
+    if (is_negative(vel_min) && is_positive(vel_max)){
+        vel_target = constrain_float(vel_target, vel_min, vel_max);
     }
 
     // velocity correction with input velocity
     vel_target += vel_input;
 
-    shape_vel_accel(vel_target, accel_input, vel, accel, vel_min, vel_max, accel_min, accel_max, tc, dt);
+    shape_vel_accel(vel_target, accel_input, vel, accel, vel_min, vel_max, accel_min, accel_max, jerk_max, dt, limit_total_accel);
 }
 
 // 2D version
 void shape_pos_vel_accel_xy(const Vector2p& pos_input, const Vector2f& vel_input, const Vector2f& accel_input,
                             const Vector2p& pos, const Vector2f& vel, Vector2f& accel,
-                            const float vel_correction_max, const float vel_max, const float accel_max, const float tc, const float dt)
+                            const float vel_max, const float accel_max,
+                            const float jerk_max, const float dt, bool limit_total_accel)
 {
-    if (!is_positive(tc)) {
+    // sanity check accel_max
+    if (!is_positive(accel_max)) {
+        INTERNAL_ERROR(AP_InternalError::error_t::invalid_arg_or_result);
         return;
     }
 
     // Calculate time constants and limits to ensure stable operation
-    const float KPv = 1.0f / (CONTROL_TIME_CONSTANT_RATIO*tc);
-    const float accel_tc_max = accel_max*(1.0f - 1.0f/CONTROL_TIME_CONSTANT_RATIO);
+    const float KPv = jerk_max / (CONTROL_TIME_CONSTANT_RATIO * accel_max);
+    const float accel_tc_max = accel_max * (1.0 - 1.0 / CONTROL_TIME_CONSTANT_RATIO);
 
     // position error to be corrected
     Vector2f pos_error = (pos_input - pos).tofloat();
 
     // velocity to correct position
     Vector2f vel_target = sqrt_controller(pos_error, KPv, accel_tc_max, dt);
 
-    // limit velocity correction to vel_correction_max
-    if (is_positive(vel_correction_max)) {
-        vel_target.limit_length(vel_correction_max);
+    // limit velocity to vel_max
+    if (is_negative(vel_max)) {
+        INTERNAL_ERROR(AP_InternalError::error_t::invalid_arg_or_result);
```

Comment:
This internal error is raised but please note that the vel_max value is still used below because there is no "return".  So this way cause a failure in SITL (which is fine) but for real boards it will just raise an internal error that will be logged and appear when the user next tries to arm.